discovery/kubernetes/client_metrics.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/metrics"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	cacheMetricsNamespace     = metricsNamesapce + "_cache"
+	workqueueMetricsNamespace = metricsNamesapce + "_workqueue"
+)
+
+var (
+	// Metrics for client-go adapters observi client's HTTP requests
+	clientGoRequestResultMetric = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamesapce,
+			Name:      "http_request_result_total",
+			Help:      "Request results of the client-go on the Kubernetes API.",
+		},
+		[]string{"status_code", "method"},
+	)
+	clientGoRequestLatencyMetric = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: metricsNamesapce,
+			Name:      "http_request_duration_seconds",
+			Help:      "Request latency of the client-go on Kubernetes API.",
+		},
+		[]string{"verb", "endpoint"},
+	)
+
+	// Definition of metrics for client-go cache metrics provider
+	clientGoCacheListTotalMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "cache_list_total",
+			Help:      "Total number of client-go Kubernets API calls.",
+		},
+	)
+	clientGoCacheListDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_duration_seconds",
+			Help:      "Duration of a Kubernetes API call in seconds.",
+		},
+	)
+	clientGoCacheItemsInListCountMetric = prometheus.NewSummary(
```

Comment:
For this metric as well as for `clientGoCacheItemsInWatchesCountMetric`, I'm not sure which metric type makes sense...

Reply:
I'm not sure either. I leave this as it is right now, but have not strong opinion on this, so if you want, I'll be happy to change it.